unified-runtime/source/adapters/level_zero/v2/queue_immediate_in_order.cpp

```diff
@@ -701,31 +703,142 @@ ur_result_t ur_queue_immediate_in_order_t::enqueueWriteHostPipe(
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY(
+      "ur_queue_immediate_in_order_t::enqueueUSMDeviceAllocExp");
+
+  return enqueueUSMAllocHelper(pPool, size, pProperties, numEventsInWaitList,
+                               phEventWaitList, ppMem, phEvent,
+                               UR_USM_TYPE_DEVICE);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMSharedAllocExp(
-    ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY(
+      "ur_queue_immediate_in_order_t::enqueueUSMSharedAllocExp");
+
+  return enqueueUSMAllocHelper(pPool, size, pProperties, numEventsInWaitList,
+                               phEventWaitList, ppMem, phEvent,
+                               UR_USM_TYPE_SHARED);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMHostAllocExp(
-    ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY("ur_queue_immediate_in_order_t::enqueueUSMHostAllocExp");
+
+  return enqueueUSMAllocHelper(pPool, size, pProperties, numEventsInWaitList,
+                               phEventWaitList, ppMem, phEvent,
+                               UR_USM_TYPE_HOST);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMFreeExp(
-    ur_usm_pool_handle_t, void *, uint32_t, const ur_event_handle_t *,
-    ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_usm_pool_handle_t pPool, void *pMem, uint32_t numEventsInWaitList,
+    const ur_event_handle_t *phEventWaitList, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY("ur_queue_immediate_in_order_t::enqueueUSMFreeExp");
+  auto commandListLocked = commandListManager.lock();
+
+  auto zeSignalEvent = getSignalEvent(commandListLocked, phEvent,
+                                      UR_COMMAND_ENQUEUE_USM_FREE_EXP);
+  auto [pWaitEvents, numWaitEvents] =
+      getWaitListView(commandListLocked, phEventWaitList, numEventsInWaitList);
+
+  umf_memory_pool_handle_t hPool = umfPoolByPtr(pMem);
+  if (!hPool) {
+    return UR_RESULT_SUCCESS;
```

Comment:
This shouldn't be a success, probably `UR_RESULT_ERROR_INVALID_MEM_OBJECT` would be appropriate.

Reply:
I was looking at `umfFree` implementation, used by `urUSMFree`:
```
umf_result_t umfFree(void *ptr) {
    umf_memory_pool_handle_t hPool = umfPoolByPtr(ptr);
    if (hPool) {
        LOG_DEBUG("calling umfPoolFree(pool=%p, ptr=%p) ...", (void *)hPool,
                  ptr);
        return umfPoolFree(hPool, ptr);
    }
    return UMF_RESULT_SUCCESS;
}
```
Here, if `hPool` is null, success is returned

Reply:
In our case, user could provide invalid pointer and we would return success. Considering current umf use-cases in L0 and L0 v2 I don't think it is possible to have a valid pointer that doesn't come from umf pool.